### PR TITLE
Feat/fix parser

### PR DESCRIPTION
**Pull request by anfibiacreativa**
## Purpose

Fix the broken parsing for the follow up questions in current version. This version does not decouple the components or introduces code optimizations. 
It minimally varies color scheme and the rendering format of follow up questions (as shown in GIF)

## Does this introduce a breaking change?

```
[ ] Yes
[ x] No
```

## Pull Request Type

What kind of change does this Pull Request introduce?

```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ x] Other... Please describe: cosmetic changes


```

## How to Test

- Get the code

```
git clone [repo-address]
cd [repo-name]
git checkout [branch-name]
npm install
```

- Test the code
No test was added

## What to Check
Follow up questions are properly rendered, as displayed in the GIF

followupquestions

### CODE REVIEW

File: packages/chat-component/src/core/parser/index.ts [hunk 1]
```diff
@@ -50,15 +52,27 @@ export async function parseStreamedMessages({
...
+      isFollowupQuestion = true;
       followupQuestionsBuffer.push(chunkValue);
       continue;
     } else if (followupQuestionsBuffer.length > 0 && chunkValue.includes('Question')) {
       isFollowupQuestion = true;
       followupQuestionsBuffer.push(chunkValue);
       continue;
+    } else if (chunkValue.includes('<<') && isFollowupQuestion) {
+      isFollowupQuestion = true;
+      chunkValue = chunkValue.replaceAll(FOLLOW_UP_QUESTION, '');
```
Comment (manekinekko) [inline packages/chat-component/src/core/parser/index.ts:69]: can you try removing this line call to see if you still have the same formatting?

File: packages/chat-component/src/core/parser/index.ts [hunk 2]
```diff
@@ -50,15 +52,27 @@ export async function parseStreamedMessages({
...
+      continue;
+    } else if (chunkValue.includes('?>') || chunkValue.includes('>')) {
+      followUpQuestionIndex = followUpQuestionIndex + 1;
+      isFollowupQuestion = true;
+      chunkValue = chunkValue.replaceAll(FOLLOW_UP_QUESTION, '');
+      continue;
     } else if (isFollowupQuestion) {
       isFollowupQuestion = true;
       chunkValue = chunkValue.replace(/:?\n/, '');
```
Comment (manekinekko) [inline packages/chat-component/src/core/parser/index.ts:78]: can you add a comment for future ref?

File: packages/chat-component/src/core/parser/index.ts [hunk 3]
```diff
@@ -50,15 +52,27 @@ export async function parseStreamedMessages({
...
       followupQuestionsBuffer.push(chunkValue);
       continue;
+    } else if (chunkValue.includes('<<') && isFollowupQuestion) {
+      isFollowupQuestion = true;
+      chunkValue = chunkValue.replaceAll(FOLLOW_UP_QUESTION, '');
+      continue;
+    } else if (chunkValue.includes('?>') || chunkValue.includes('>')) {
+      followUpQuestionIndex = followUpQuestionIndex + 1;
+      isFollowupQuestion = true;
+      chunkValue = chunkValue.replaceAll(FOLLOW_UP_QUESTION, '');
```
Comment (manekinekko) [inline packages/chat-component/src/core/parser/index.ts:74]: can you try removing this line call to see if you still have the same formatting?

**Review by sinedied — COMMENTED** [review]
Does it work when it's not streaming?
As I recall there is a separate path for non-streaming mode

File: packages/chat-component/src/core/parser/index.ts (hunk 2 above)
Comment (anfibiacreativa) [reply packages/chat-component/src/core/parser/index.ts:78]: About the particular logic in this line? Or in general?

File: packages/chat-component/src/core/parser/index.ts (hunk 1 above)
Comment (anfibiacreativa) [reply packages/chat-component/src/core/parser/index.ts]: Done.

File: packages/chat-component/src/core/parser/index.ts (hunk 2 above)
Comment (manekinekko) [reply packages/chat-component/src/core/parser/index.ts:75]: I'd say in general, to keep track of some of the decisions made.

**Comment by anfibiacreativa** [comment]
> Does it work when it's not streaming? As I recall there is a separate path for non-streaming mode

Yes! Uploading gif to teams since it's too large.

File: packages/chat-component/src/core/parser/index.ts [hunk 4]
```diff
@@ -52,9 +52,10 @@ export async function parseStreamedMessages({
     // followup questions are marked either with the word 'Next Questions:' or '<<text>>' or both at the same time
     // these markers may be split across multiple chunks, so we need to buffer them!
     // TODO: remove all this logic from the frontend and implement a solution on the backend or with TypeChat
-
+    // we start by creating a buffer when we match the first marker
     const matchedFollowupQuestionMarker =
       (!isFollowupQuestion && chunkValue.includes('Next')) || chunkValue.includes('<<');
```
Comment (manekinekko) [inline packages/chat-component/src/core/parser/index.ts:57]: btw should we still parse `Next` markers? Is the new model still using `Next` to format followup questions?

Comment (manekinekko) [reply packages/chat-component/src/core/parser/index.ts:57]: If the updated model uses only `<<` and `>>` to format followup questions, then I'd suggest changing the condition to:

`const matchedFollowupQuestionMarker = !isFollowupQuestion && chunkValue.includes('<<');`

File: packages/chat-component/src/core/parser/index.ts [hunk 5]
```diff
@@ -63,13 +64,18 @@ export async function parseStreamedMessages({
...
       followupQuestionsBuffer.push(chunkValue);
       continue;
+      // if we're already processing questions, we don't need to check for the marker again
+      // but we need to check if we reached the end of the followup questions
     } else if (chunkValue.includes('<<') && isFollowupQuestion) {
       isFollowupQuestion = true;
       continue;
+      // this updates the index, so we add each question to a different array entry
+      // to simplify styling
     } else if (chunkValue.includes('?>') || chunkValue.includes('>')) {
```
Comment (manekinekko) [inline packages/chat-component/src/core/parser/index.ts:74]: why are we checking for `?>` && `>` and not `?>>` && `>>`?

**Comment by manekinekko** [comment]
Love the comments, thank you!

File: packages/chat-component/src/core/parser/index.ts (hunk 4 above)
Comment (anfibiacreativa) [reply packages/chat-component/src/core/parser/index.ts:57]: Good question. @sinedied  can you clarify?

File: packages/chat-component/src/core/parser/index.ts (hunk 5 above)
Comment (anfibiacreativa) [reply packages/chat-component/src/core/parser/index.ts:74]: Because in all tests I ran, ?> and the last > were returned as part of 2 different chunks.

**Pull request merged**